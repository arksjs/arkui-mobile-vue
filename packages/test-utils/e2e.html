<script src="../../node_modules/vue/dist/vue.global.js"></script>
<script src="../../dist/index.js"></script>
<link rel="stylesheet" href="../../dist/index.css" />

<div id="app"></div>
