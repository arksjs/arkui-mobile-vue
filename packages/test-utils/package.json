{
  "name": "@arksjs/test-utils",
  "version": "0.0.1",
  "description": "arkui test-utils",
  "main": "dist/index.js",
  "private": true
}
