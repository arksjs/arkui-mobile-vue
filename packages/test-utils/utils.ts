export const timeout = (n: number) => new Promise(r => setTimeout(r, n))
