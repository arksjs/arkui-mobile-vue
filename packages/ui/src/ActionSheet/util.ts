import { cloneData, isObject } from '../helpers'
import type { Option } from './types'

export const getOptions = (options?: Option[]) => {
  const newOptions: Option[] = []

  if (Array.isArray(options)) {
    options.forEach(v => {
      newOptions.push(
        isObject(v)
          ? cloneData(v)
          : {
              name: v.toString()
            }
      )
    })
  }

  return newOptions
}

export const getItemClasses = (option: Option) => {
  return [
    'ta-action-sheet_item',
    'ta-horizontal-hairline',
    { disabled: !!option.disabled, highlight: !!option.highlight }
  ]
}
