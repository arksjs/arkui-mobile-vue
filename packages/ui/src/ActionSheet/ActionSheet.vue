<template>
  <Drawer
    class="ta-action-sheet"
    :title="title"
    placement="bottom"
    :visible="visible"
    @visibleStateChange="onVisibleStateChange"
    @confirm="onConfirm"
    @cancel="onCancel"
    @update:visible="onUpdateVisible"
    ref="popupRef"
  >
    <ul class="ta-action-sheet_list">
      <li
        :class="getItemClasses(item)"
        v-for="(item, index) in options2"
        :key="index"
        @click="onItemClick(index)"
      >
        <div class="ta-action-sheet_item-inner">
          <span>{{ item.name }}</span>
          <small v-if="item.description">{{ item.description }}</small>
        </div>
      </li>
    </ul>
    <ul class="ta-action-sheet_list" v-if="showCancel">
      <li class="ta-action-sheet_item ta-horizontal-hairline" @click="onCancelClick">
        <div class="ta-action-sheet_item-inner align--center">
          <span>{{ cancelText || locale.actionSheetCancelText }}</span>
        </div>
      </li>
    </ul>
  </Drawer>
</template>

<script lang="ts">
import { computed, defineComponent, type PropType } from 'vue'
import { Drawer } from '../Drawer'
import { usePopupExtend } from '../popup/use-popup'
import { popupEmits, popupExtendProps } from '../popup/props'
import type { Option, Detail, OnConfirm, ActionSheetEmits } from './types'
import { useLocale } from '../ConfigProvider/context'
import { isNumber, isString, type PropsToEmits, type VoidFnToBooleanFn } from '../helpers'
import { getItemClasses, getOptions } from './util'

const confirmValidator: VoidFnToBooleanFn<OnConfirm> = payload =>
  payload && isNumber(payload.index) && payload.item && isString(payload.item.name)

export default defineComponent({
  name: 'ta-action-sheet',
  components: { Drawer },
  props: {
    ...popupExtendProps,
    title: {
      type: String
    },
    showCancel: {
      type: Boolean,
      default: false
    },
    cancelText: {
      type: String
    },
    options: {
      type: Array as PropType<Option[]>,
      default: () => [] as Option[]
    }
  },
  emits: {
    ...popupEmits,
    confirm: confirmValidator
  } as PropsToEmits<ActionSheetEmits>,
  setup(props, ctx) {
    const { locale } = useLocale()
    const popup = usePopupExtend<Detail>(ctx)

    function onItemClick(index: number) {
      popup.customConfirm({
        index,
        item: {
          name: props.options[index].name
        }
      })
    }

    const options2 = computed(() => getOptions(props.options))

    return {
      ...popup,
      options2,
      onItemClick,
      locale,
      getItemClasses
    }
  }
})
</script>
