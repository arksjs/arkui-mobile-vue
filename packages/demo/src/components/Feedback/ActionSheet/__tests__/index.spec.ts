import { shallowMount } from '@vue/test-utils'
import { ActionSheet } from '@/ActionSheet'

const options = [
  {
    name: 'Option 1'
  },
  {
    name: 'Option 2'
  },
  {
    name: 'Option 3'
  }
]

describe('ActionSheet', () => {
  test('should render prop set options correctly', () => {
    const wrapper = shallowMount(ActionSheet, {
      props: {
        options
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should render props showCancel=true & cancelText="Custom Cancel" correctly', () => {
    const wrapper = shallowMount(ActionSheet, {
      props: {
        showCancel: true,
        cancelText: 'Custom Cancel'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
