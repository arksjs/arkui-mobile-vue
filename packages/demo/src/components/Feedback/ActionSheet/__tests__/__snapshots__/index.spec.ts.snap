// Vitest Snapshot v1

exports[`ActionSheet > should render prop set options correctly 1`] = `"<drawer-stub visible=\\"false\\" maskclosable=\\"true\\" placement=\\"bottom\\" showclose=\\"false\\" enablesafeareainsets=\\"true\\" showmask=\\"true\\" class=\\"fx-action-sheet\\"></drawer-stub>"`;

exports[`ActionSheet > should render props showCancel=true & cancelText="Custom Cancel" correctly 1`] = `"<drawer-stub visible=\\"false\\" maskclosable=\\"true\\" placement=\\"bottom\\" showclose=\\"false\\" enablesafeareainsets=\\"true\\" showmask=\\"true\\" class=\\"fx-action-sheet\\"></drawer-stub>"`;
